### PR TITLE
New category: Onions

### PR DESCRIPTION
**Pull request by MarcAbonce**
With engines Ahmia, Torch and Not Evil.

Disabled by default. Can be enabled by setting the SOCKS proxies to wherever Tor is listening and setting using_tor_proxy as True.

To avoid manually adding the timeout on each engine, you can set extra_proxy_timeout to account for Tor's (or whatever proxy used) extra time.

Should solve #413 and probably #519

### CODE REVIEW

**Comment by GreenLunar** [comment]
There is also **not Evil** at http://hss3uro2hsxfogfq.onion/ (was **TorSearch** at http://kbhpodhnfxl3clb4.onion/)

**Comment by MarcAbonce** [comment]
By the way, ~~htk2rvtgj6vv3ppj.onion~~ is currently running this branch.
Edit: ~~mze6ter6fchwojcr.onion~~ ~~http://jiwfbtg2kfkadjku.onion~~ http://7ea7vegvcqgdzxjz.onion

**Comment by asciimoo** [comment]
@a01200356 great, the only problem is that it isn't compatible with the existing non-tor configuration. It would be cool to provide both to the users. Could be a separated `settings_tor.yml` for the tor config.

**Comment by MarcAbonce** [comment]
Sorry for the delay.
I just updated the PR. But instead of having two settings files, I added the using_tor_proxy setting that should automatically set up all the other Tor specific configurations.
When set to False or commented out, as default, everything should behave exactly the same as current version on upstream.

**Comment by gsantner** [comment]
@a01200356 I'm currently setting up searx and gonna try the version of your branch.
I can give some feedback If I run into problems.

**Comment by MarcAbonce** [comment]
I just fixed the issue I had with not evil's encoding.

About the last commit, I just assume that using Tor Browser's user agent would make our traffic seem more "normal" when using Tor than using a random user agent. However, this is merely an assumption on my part. I can't really tell if it actually helps or not.

**Comment by ghost** [comment]
Looks interesting.

File: searx/utils.py [hunk 1]
```diff
@@ -33,14 +33,18 @@
...
 
 blocked_tags = ('script',
                 'style')
 
 
 def gen_useragent():
     # TODO
-    return ua.format(os=choice(ua_os), version=choice(ua_versions))
+    if settings['outgoing'].get('using_tor_proxy'):
+        return tor_browser_ua
```
Comment (dalf) [inline searx/utils.py]: Why a specific behavior for Tor ?

**Comment by MarcAbonce** [comment]
~~With the fix in #1121, Google doesn't block Tor exit nodes any more.~~
I also reverted the Tor Browser user-agent in Tor requests, I don't think it makes any difference anyway.
Also, DuckDuckGo currently uses the normal website because the hidden service is extremely slow, which makes it reach timeout most of the time.
An updated instance running this branch is in ~~http://jiwfbtg2kfkadjku.onion~~ http://7ea7vegvcqgdzxjz.onion

**Comment by asciimoo** [comment]
@MarcAbonce sorry for the delayed answer. I still think that this is a useful functionality, could you please rebase it to the latest master? I'm happy to do it if you don't have capacity.

**Comment by MarcAbonce** [comment]
I rebased to master and I also added another commit.

**Review by asciimoo — APPROVED** [review]
Awesome, thanks!

**Comment by asciimoo** [comment]
This is a great addition, we should mention it in the documentation.

**Comment by dalf** [comment]
Note about the plugin, it can't be disabled, I guess on purpose :
https://github.com/searx/searx/blob/db703a0283ee169381aeea97c678e666ae508348/searx/plugins/__init__.py#L179-L188

The plugin is loaded after `enabled_plugins` parsing and in the preferences (in the oscar theme), it is hidden:
https://github.com/searx/searx/blob/db703a0283ee169381aeea97c678e666ae508348/searx/templates/oscar/preferences.html#L261

**Comment by MarcAbonce** [comment]
Yeah, maybe it would be better to extend the new preferences lock to include plugins and use that instead?

**Comment by Perflyst** [comment]
Looks like if a proxy is configured the proxy is used for all search engines and not only for onion search engines. Is there an option to use the proxy (in this specific use-case the Tor socks proxy) only for onion search engines?

**Comment by dalf** [comment]
You can specify a proxy per engine.
See https://searx.github.io/searx/admin/settings.html#engine-settings

So you have to copy / paste the proxy section for each onion engine
or you can use YAML anchor ( https://en.wikipedia.org/wiki/YAML#Advanced_components )

**Pull request merged**